protocols/request-response/src/lib.rs

```diff
@@ -382,26 +374,27 @@ where
                 peer_id: peer.clone(),
                 condition: DialPeerCondition::Disconnected,
             });
-            self.pending_requests.entry(peer.clone()).or_default().push(request);
+            self.pending_outbound_requests.entry(peer.clone()).or_default().push(request);
         }
 
         request_id
     }
 
     /// Initiates sending a response to an inbound request.
     ///
-    /// If the `ResponseChannel` is already closed due to a timeout or
-    /// the connection being closed, the response is returned as an `Err`
-    /// for further handling. Once the response has been successfully sent
-    /// on the corresponding connection, [`RequestResponseEvent::ResponseSent`]
-    /// is emitted.
+    /// Once the response has been successfully sent on the corresponding
+    /// connection, [`RequestResponseEvent::ResponseSent`] is emitted.
     ///
     /// The provided `ResponseChannel` is obtained from an inbound
     /// [`RequestResponseMessage::Request`].
-    pub fn send_response(&mut self, ch: ResponseChannel<TCodec::Response>, rs: TCodec::Response)
-        -> Result<(), TCodec::Response>
```

Comment:
There is no harm in keeping this signature, is there? It can still be useful to get early feedback here (and to get the response back).

Reply:
> There is no harm in keeping this signature, is there?

True. At the same time I find the error handling of `libp2p-request-response` with `send_response` returning a `Result` to be quite complex. E.g. when `send_response` returns `Ok(())`, one can not assume that sending the response succeeded, but instead only assume that the corresponding connection upgrade is still alive. While one can document this edge case I would prefer usage of `libp2p-request-response` not to require any knowledge about it internals. By not returning a `Result` there is only a single way of error reporting i.e. through the events emitted by the behaviour.

> It can still be useful to get early feedback here (and to get the response back).

To be honest I did not properly consider this use-case. Thinking about it some more I don't think returning the failed response payload would be helpful for Substrate. If a response fails, I would expect the remote to initiate a new request and I think keeping the response around until the remote retries by sending a new request introduces unnecessary complexity. This of course is just one data point. Can you think of a use-case where keeping the response around would be helpful?

The above said, this is not a strong opinion. Off the top of my head I would favor simplicity over complexity, especially as one can still change `send_response` to return a `Result` later on.

What do you think in regards to the above reasoning @romanb?

Reply:
>  Can you think of a use-case where keeping the response around would be helpful?

Any situation where computing parts of the response may have been expensive and one may want to store or send these parts to some other peer, without having to reconstruct them. This may be contrived, it just seems to be a shame to throw away information (both the response and the knowledge about whether the inbound request already failed) at a point where it could be given. Note that initially, I actually did the same thing you've done here, likely for very similar reasons. The `Result` return value is a rather recent addition of mine which would be reverted again here. I am fine either way, I don't think I can make a strong case for keeping it at the moment.

Reply:
https://github.com/libp2p/rust-libp2p/pull/1886/commits/324b2d7f9779026840d84d8023f19a621829d4dd reverts the change to the signature of `send_response` and in the doc comment stresses the point that a response can still fail after `send_response` returned `Ok`.

Thanks for bringing this up!